core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -741,7 +741,9 @@ public boolean append(Object kbase, long koff, int klen, Object vbase, long voff
         longArray.set(pos * 2 + 1, keyHashcode);
         isDefined = true;
 
-        if (numKeys >= growthThreshold && longArray.size() < MAX_CAPACITY) {
```

Comment:
can you remind me some more details of `BytesToBytesMap`? What happens if we don't grow? I don't see a loop in this method and not sure how the job hangs.

Reply:
Ah, sure.

The client of `BytesToBytesMap`, like HashAggregate, will call `lookup` to find a `Location` to write value. The returned location will be used to do append (`Location.append`).  Everytime after we append a key/value, we check if it is time to grow internal array and grow up it if needed.

`lookup` delegates looking up keys to `safeLookup`. Its control flow looks like:

```java
int pos = hash & mask;
int step = 1;
// an infinite loop until find matching key or empty slot.
while (true) {
  if (empty slot found) {
    ...
  } else {
    // check if matching key
    ...
  }
  // increase pos and step
  pos = (pos + step) & mask;
  step++;
}
```

So the job hangs in this loop because it can not find any empty location as the internal array is full.

We early stop growing the internal array due to wrongly check array size at:

```java
if (numKeys >= growthThreshold && longArray.size() < MAX_CAPACITY) {
  ...
}
```

Another point https://github.com/apache/spark/pull/26828#discussion_r356370350 is we may want to set `canGrowArray` to false once we are close to max capacity, so we can avoid infinite loop again.

Reply:
shouldn't `lookup` throw OOM if no space can be found?

Reply:
`lookup` just looks up empty slot in the internal array for a new key. It does not allocate memory. The array is allocated/grown up in last time `append`.

Once an empty slot (a `Location` object) is found, the client of `BytesToBytesMap` may call `append` to the Location, OOM could be thrown during append key/value.

Reply:
let me confirm the problem: so `append` mistakenly think there is enough space, and doesn't grow the array. This makes the client of `BytesToBytesMap` keeping calling `lookup` and hand. Is my understanding correct?

Reply:
`append` mistakenly think it can not grow the array anymore so does not grow the array. It keeps append value until full. Then the client calling `lookup` can not find an empty slot and gets stuck in infinite loop.

Reply:
I see, the `append` doesn't grow the array while it should. This makes `BytesToBytesMap` malformed (the array is not big enough to serve the data region) and causes problems.